inverse/private/ft_inv.m

```diff
@@ -0,0 +1,144 @@
+%               at which the singular value spectrum will be clipped. The
+%               default value is 10*eps*max(size(A))
+% kappa     = scalar integer, reflects the ordinal singular value at which
+%               the singular value will be clipped.
+% lambda    = scalar, or string (expressed as a percentage), specifying the
+%               regularization parameter for diagonal loading. Lambda 
+%               specified as a percentage will be converted into a
+%               percentage of the average of trace(A).
+% feedback  = boolean, false or true, to show the singular value spectrum
+%               with the clipping level used.
+% interactive = boolean, false or true, to manually specify a value for
+%               kappa.
+%
+% kappa and tolerance are mutually exclusive, in case both are specified,
+% tolerance takes precedence.
+
+method    = ft_getopt(varargin, 'method', 'pinv');
+tolerance = ft_getopt(varargin, 'tolerance', []);
+kappa     = ft_getopt(varargin, 'kappa', []);
+lambda    = ft_getopt(varargin, 'lambda', 0);
+feedback  = istrue(ft_getopt(varargin, 'feedback', false));
+interactive = istrue(ft_getopt(varargin, 'interactive', false));
+
+[m, n] = size(A);
+
+if n < m
+  % recurse into ft_inv on the transposed input
+  Ainv = ft_inv(A',varargin{:})';
+  return
+elseif m <= n 
+  % here the work starts
+  if ~isempty(kappa) && kappa > m
+    ft_error('the kappa value for truncation is larger than the dimensionality of the matrix');
+  end
+  
+  if ~isempty(kappa) && ~isempty(tolerance)
+    ft_warning('both a kappa value and tolerance value are defined, using the kappa value for truncation');
+  end
+  
+  if isempty(lambda)
+    lambda = 0;
+  elseif ~isempty(lambda) && ischar(lambda) && lambda(end)=='%'
+    ratio  = sscanf(lambda, '%f%%');
+    ratio  = ratio/100;
+    lambda = ratio * trace(A)./m;
+  end
+  
+  if lambda~=0 && m~=n
+    ft_error('with lambda regularization, the input matrix should be square');
+  end
+  
+  if m==n && lambda~=0
+    % regularize
+    A = A + eye(m).*lambda;
+  elseif lambda~=0
+    ft_error('regularization is not possible when the matrix is not square');
+  end
+  
+  [U,S,V] = svd(A,0);
+  if m > 1
+    s = diag(S);
+  elseif m == 1
+    s = S(1);
+  else
+    s = 0;
+  end
+  
+  if isempty(kappa) && isempty(tolerance) && interactive
+    figure; plot(1:m, log10(s),'o-');
+    kappa = input('Specify the number of dimensions at which the eigenvalue spectrum will be truncated: ');
+    if isempty(kappa)
+      kappa = m;
+    end    
+  elseif isempty(kappa) && isempty(tolerance)
+    % put the tolerance for the pinv to its (increased by R.O.) default 
+    tolerance = 10 * max(m,n) * eps;
+  end
+  
+  if ~isempty(kappa)
+    % use kappa
+    usekappa = true;
+    usetolerance = false;
+  elseif ~isempty(tolerance)
+    usetolerance = true; % needed for winsorization
+    usekappa = false;
+    kappa = sum(s./s(1) > tolerance);
+  end
+  
+  if kappa==0
```

Comment:
I think it would be good to have this (and/or similar) exceptions further to the top. E.g. empty matrix, 1x1 input matrix, tolerance=inf, etc.